core/include/cond.h

```diff
@@ -0,0 +1,223 @@
+ * the reading thread blocks until it is not empty. If the pipe is closed by
+ * the sender, waiting reading threads wake up.
+ *
+ * Here is a sketch of how to implement such a structure with condition
+ * variables. For simplicity, messages are single bytes. We assume a FIFO data
+ * structure queue_t. We assume it is unsafe to add to the queue if it is full,
+ * or remove from the queue if it is empty.
+ *
+ * typedef struct pipe {
+ *     queue_t queue;
+ *     cond_t read_cond;
+ *     cond_t write_cond;
+ *     mutex_t lock;
+ *     bool closed;
+ * } pipe_t;
+ *
+ * void pipe_init(pipe_t* pipe) {
+ *     queue_init(&pipe->queue);
+ *     cond_init(&pipe->read_cond);
+ *     cond_init(&pipe->write_cond);
+ *     mutex_init(&pipe->lock);
+ *     pipe->closed = false;
+ * }
+ *
+ * void pipe_write(pipe_t* pipe, char c) {
+ *     mutex_lock(&pipe->lock);
+ *     while (queue_length(&pipe->queue) == MAX_QUEUE_LENGTH && !pipe->closed) {
+ *         cond_wait(&pipe->write_cond, &pipe->lock);
+ *     }
+ *     if (pipe->closed) {
+ *         mutex_unlock(&pipe->lock);
+ *         return 0;
+ *     }
+ *     add_to_queue(&pipe->queue, c);
+ *     cond_signal(&pipe->read_cond);
+ *     mutex_unlock(&pipe->lock);
+ *     return 1;
+ * }
+ *
+ * void pipe_close(pipe_t* pipe) {
+ *     mutex_lock(&pipe->lock);
+ *     pipe->closed = true;
+ *     cond_broadcast(&pipe->read_cond);
+ *     cond_broadcast(&pipe->write_cond);
+ *     mutex_unlock(&pipe->lock);
+ * }
+ *
+ * int pipe_read(pipe_t* pipe, char* buf) {
+ *     mutex_lock(&pipe->lock);
+ *     while (queue_length(&pipe->queue) == 0 && !pipe->closed) {
+ *         cond_wait(&pipe->read_cond, &pipe->lock);
+ *     }
+ *     if (pipe->closed) {
+ *         mutex_unlock(&pipe->lock);
+ *         return 0;
+ *     }
+ *     *buf = remove_from_queue(&pipe->queue);
+ *     cond_signal(&pipe->write_cond);
+ *     mutex_unlock(&pipe->lock);
+ *     return 1;
+ * }
+ *
+ * Note that this could actually be written with a single condition variable.
+ * However, the example includes two for didactic reasons.
+ *
+ * @author      Sam Kumar <samkumar@berkeley.edu>
+ */
+
+#ifndef COND_H
+#define COND_H
+
+#include <stdbool.h>
+#include <stddef.h>
+
+#include "list.h"
+#include "mutex.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Condition variable structure. Must never be modified by the user.
+ */
+typedef struct {
+    /**
+     * @brief   The process waiting queue of the condition variable.
+     * @internal
+     */
+    list_node_t queue;
+} cond_t;
+
+/**
+ * @brief   Static initializer for cond_t.
+ * @details This initializer is preferable to cond_init().
+ */
+#define COND_INIT { { NULL } }
+
+/**
+ * @brief Initializes a condition variable.
+ * @details For initialization of variables use COND_INIT instead.
+ *          Only use the function call for dynamically allocated condition
+ *          variables.
+ * @param[out] cond    pre-allocated condition structure, must not be NULL.
```

Comment:
looks more like an `[in]` param to me, or not? At least the object/memory is not allocated inside that function and passed out, it just modifies attributes of the `cond_t` passed in.

Reply:
Done. I was originally trying to be consistent with the comment for `mutex_init` in mutex.h.